<a href='http://www.algebra.hr'> <img src='../algebra_logo_color_h.png' alt="Algebra" width="500" /></a>
___
# Pandas manipulacija podacima

Do sada smo naučili kako učitati podatke i kako ih prilagoditi za daljnju obradu. Pokušajmonešto pročitati iz tih podataka. Koristit ćemo se podacima iz csv datoteka:<br>
- *Data\Vjezbe_Primjeri\airbnb_new_york_city_listings.csv* - sažetak podataka i mjerenja za ponudu na AirBnB u New Yorku
- *Data\Vjezbe_Primjeri\airbnb_new_york_city_reviews_full.csv* - podaci o ocjenama ponude na AirBnB u New Yorku
- *Data\Vjezbe_Primjeri\mall_customers.csv* - za potrebe dodatnog pojašnjavanja funkcionalnosti Pandas biblioteke - podaci su preuzeti s https://www.kaggle.com/kandij/mall-customers
- *Data\Vjezbe_Primjeri\ocjene.csv* - za potrebe dodatnog pojašnjavanja funkcionalnosti Pandas biblioteke - podaci su generirani nasumično pomoću web usluge https://www.json-generator.com/ preuzete s http://insideairbnb.com/get-the-data.html.

Za početak, pripremimo ih za analizu što ćemo iskoristiti kao malo ponavljanje prvog dijela gradiva o Pandas biblioteci.

In [168]:
import pandas as pd

listings_df = pd.read_csv('../Data/Vjezbe_Primjeri/airbnb_new_york_city_listings.csv')
# reviews_df = pd.read_csv('../Data/Vjezbe_Primjeri/airbnb_new_york_city_reviews_full.csv')
# mall_customers_df = pd.read_csv('../Data/Vjezbe_Primjeri/mall_customers.csv')
# ocjene_df = pd.read_csv('../Data/Vjezbe_Primjeri/ocjene.csv')
# Učitat ćemo kasnije ovisno o potrebi.

In [169]:
listings_df # Isti rezultat mozete dobiti ako koristite metodu display() -> display(listings_df)

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                   Large Furnished Room Near B'way　      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt    150              30                 48   
1     -73.957650  Entire home/apt     75               1                408   
2     -73.955120     Private room     60              30                 50   
3     -73.994540  Entire home/apt    275               5                  1   
4     -73.983170     Private room     61               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt    255              31                  0   
36720 -73.969239  Entire home/apt    306               5                  0   
36721 -73.960455  Entire home/apt     63              30                  0   
36722 -73.797953     Private room     75               1                  0   
36723 -73.981209  Entire home/apt    115              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719         NaN                NaN                               6   
36720         NaN                NaN                               1   
36721         NaN                NaN                              96   
36722         NaN                NaN                               1   
36723         NaN                NaN                              11   

       availability_365  
0                   341  
1                   212  
2                   365  
3                   184  
4                   255  
...                 ...  
36719               364  
36720                16  
36721               334  
36722                87  
36723               364  

[36724 rows 

Prvo ćemo se upoznati s tipovima podataka u kolonama. Oni će nam pomoći u slučaju kada neki tip podatka trebamo promijeniti i kada želimo popuniti prazne ćelije podacima. Primjer je kolona "Name", koja je tekstualnog tipa i svako ćemo prazno polje popuniti, tako što ćemo upisati vrijednost "Nedostaje ime", a za numeričke tipove ćemo dodati vrijednost 0.00 ili samo 0. Sve navedeno čini veliku razliku.

In [170]:
listings_df.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

Kolona "last_review" je object tip podataka, a u toj koloni su datumi zadnjih ocjena. Promijenimo tip podataka u toj koloni, u *DateTime* tipu podatka, pomoću metode *to_datetime()* koja kao argument prima naziv kolone.

*INFORMACIJA: mogli bismo koristiti i naš "stari" način pomoću astype() metode kojoj kao argument predamo naziv tipa podatka. Primjer prikazan unutar komentara.*

Kolona "price" je cjelobrojnog tipa podataka, a prikladniji je decimalni pa promijenimo tip i toj koloni.

In [171]:
# listings_df['last_review'] = listings_df['last_review'].astype('datetime64')

listings_df['last_review'] = pd.to_datetime(listings_df['last_review'])
listings_df['price'] = listings_df['price'].astype('float')
listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
dtype: object

Sada kada imamo tip podatka *DateTime*, to možemo iskoristiti za neke dodatne akcije nad DF-om, kao što su na primjer dodavanje nove kolone naziva "last_review_year" u kojoj će biti pohranjeni podaci o godini zadnje ocjene, ali ne i cijeli datum.

In [172]:
listings_df['last_review_year'] = listings_df['last_review'].dt.year
listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                   Large Furnished Room Near B'way　      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719         NaT                NaN                               6   
36720         NaT                NaN                               1   
36721         NaT                NaN                              96   
36722         NaT                NaN                               1   
36723         NaT                NaN                              11   

       availability_365  last_review_year  
0                   341            2019.0  
1                   212            2021.0  
2                   365            2016.0  
3                   184            2014.0  
4                   255            2021.0  
...                 ...               ...  
36719            

Sada imamo novu kolonu, ali su u njoj pohranjeni podaci u decimalnom zapisu. Godine trebaju biti zapisane kao cijeli brojevi. To ćemo riješiti nakon provjere toga  imamo li prazne ćelije.

Mi smo koristili *empty* svojstvo, ali ćemo pomoću metode dobiti malo više podataka o praznim ćelijama.

In [173]:
listings_df.isnull().sum()

id                                   0
name                                13
host_id                              0
host_name                           23
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            0
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       9415
reviews_per_month                 9415
calculated_host_listings_count       0
availability_365                     0
last_review_year                  9415
dtype: int64

In [174]:
listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                         float64
dtype: object

Imamo pet kolona koje imaju prazne ćelije. Kolone "name" i "host_name" su tipa object, odnosno tekstualnog tipa, dok su "reviews_per_month" i "last_review_year" decimalni brojevi, a kolona "last_review" je tipa *DateTime*.

Unesimo podatke u prazne ćelije tako da nam kasnije ne utječu na proračune. To će najbolje biti ako za kolonu:<br>
- "name" unesemo vrijednost "No Name Listing"
- "host_name" unesemo vrijednost "Bez imena" ili engleski uvriježeni oblik za bezimene "Jane Doe" za ženske osobe ili "John Doe" za muške osobe, nadalje mi ćemo koristi "J. Doe" da ostanemo neutralni
- "reviews_per_month" unesemo vrijednost 0.00 jer broj ocjena mjesečno može biti decimalni broj
- "last_review_year" unesemo vrijednost 0
- "last_review" unesemo vrijednost '01-01-1970' što bi odgovaralo početku UNIX vremena

Za ovu aktivnost imamo na raspolaganju tri metode:<br>
- *fillna()*
- *replace()*
- *interpolate()*

Mi ćemo koristiti metodu *fillna()*.

*INFORMACIJA: više detalja o navedenim funkcijama možete pronaći na stranici: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html.*

In [175]:
listings_df['name'].fillna('No Name Listing', inplace=True)
listings_df['host_name'].fillna('J. Doe', inplace=True)
listings_df['reviews_per_month'].fillna(0.00, inplace=True)
listings_df['last_review_year'].fillna(0.00, inplace=True)
listings_df['last_review'].fillna('01-01-1970', inplace=True)

Provjerimo ponovno koliko je ćelija koje nemaju vrijednost.

In [176]:
listings_df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
last_review_year                  0
dtype: int64

Odlično, nema niti jedna prazna ćelija. Provjerimo opet tipove podataka jer je moguće da smo ovom promjenom napravili i neke druge koje nismo željeli.

In [177]:
listings_df.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                             float64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
last_review_year                  float64
dtype: object

Kolona "last_review" ponovno je tekstualni tip podatka, a "last_review_year" kolonu smo ionako željeli prebaciti u cjelobrojni tip umjesto decimalnog. Napravimo to za ove dvije kolone.

In [178]:
listings_df['last_review_year'] = listings_df['last_review_year'].astype('int64')
listings_df['last_review'] = pd.to_datetime(listings_df['last_review'])
listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                           int64
dtype: object

Sada je već bolje. Prikažimo naš DF da vidimo što smo napravili.

In [179]:
listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                   Large Furnished Room Near B'way　      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  
0                   341              2019  
1                   212              2021  
2                   365              2016  
3                   184              2014  
4                   255              2021  
...                 ...               ...  
36719            

Odlično, imamo jednu kolonu više od inicijalnog skupa podataka, nemamo praznih ćelija.

Većinu tipova podataka smo očistili, međutim, često se dogodi da nam tekstualni podaci sadrže znakove koje računala ne prikazuju (primjer je znak za novi red '*\n*'). Ti znakovi su tu, no kako nam ne trebaju u podacima, bilo bi ih dobro obrisati. Pandas podržava string metodu *str.strip()* nad *DataFrame* objektom. Ova metoda briše sve posebne znakove, prazna mjesta (razmaknicu ili *space*) te slične znakove na početku i na kraju tekstualnog podatka.

Očistimo i tekstualne podatke pomoću *str.strip()* metode, u svakoj koloni koja ima tekstualni tip podatka.

*INFORMACIJA: možemo koristiti i str.lower() i str.upper() metode ako nam trebaju. U našem slučaju one nisu korisne. Više informacija možete pronaći na stranici: https://pandas.pydata.org/docs/reference/api/pandas.Series.str.strip.html.*

In [180]:
listings_df['name'] = listings_df['name'].str.strip()
listings_df['host_name'] = listings_df['host_name'].str.strip()
listings_df['neighbourhood_group'] = listings_df['neighbourhood_group'].str.strip()
listings_df['neighbourhood'] = listings_df['neighbourhood'].str.strip()
listings_df['room_type'] = listings_df['room_type'].str.strip()

In [181]:
listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                    Large Furnished Room Near B'way      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  
0                   341              2019  
1                   212              2021  
2                   365              2016  
3                   184              2014  
4                   255              2021  
...                 ...               ...  
36719            

Provjerimo jesmo li napravili neke neplanirane promjene.

In [182]:
listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                           int64
dtype: object

In [183]:
listings_df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
last_review_year                  0
dtype: int64

Napravimo još jednu preinaku u podacima kako bismo krenuli s uporabom podataka. Za početak prikažimo još jednom sve kolone:

In [184]:
listings_df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365', 'last_review_year'],
      dtype='object')

Vidimo da imamo kolone naziva "price" i "minimum_nights". Jedna kolona prikazuje cijenu najma po noćenju, a druga prikazuje koliko se minimalno noćenja mora uzeti. Ako spojimo ove dvije kolone, možemo dobiti novu koja će nam prikazati podatak o minimalnom prihodu. Dakle, ako pomnožimo cijenu s minimalnim brojem noćenja, onda je to minimalni prihod koji iznajmljivač može dobiti za svoju nekretninu.

Ako nova kolona prikazuje podatak o minimalnom prihodu, onda ćemo je nazvati "minimal_revenue". Napravimo to.

In [186]:
listings_df['minimal_revenue'] = listings_df['price'] * listings_df['minimum_nights']
listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                    Large Furnished Room Near B'way      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  minimal_revenue  
0                   341              2019           4500.0  
1                   212              2021             75.0  
2                   365              2016           1800.0  
3                   184              2014           1375.0  
4                   

In [187]:
listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                           int64
minimal_revenue                          float64
dtype: object

In [188]:
listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                    Large Furnished Room Near B'way      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  minimal_revenue  
0                   341              2019           4500.0  
1                   212              2021             75.0  
2                   365              2016           1800.0  
3                   184              2014           1375.0  
4                   

Sada trebamo sve ovo ponoviti za skup podataka o ocjenama nekretnina, za skup podataka ocjena korisnika trgovačkog lanca i ocjene studenata.

Krenimo s ocjenama nekretnina. Ovaj skup podataka ima kolonu s datumima pa ćemo ovaj put konverziju datuma napraviti prilikom učitavanja podataka izravno u *read_csv()* metodi, koja prima dva argumenta vezana uz konverziju datuma. To su:<br>
- *parse_dates*
- *date_parser.*

Više detalja na: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html.

Mi ćemo koristiti *parse_dates* argument tako što ćemo tom argumentu proslijediti listu sa samo jednim elementom, odnosno nazivom kolone.

**VJEŽBA**<br>
**Probajte ovu opciju tako da argumentu parse_dates proslijedite Boolean vrijednost *True parse_dates=True.***

In [189]:
#reviews_df = pd.read_csv('../Data/Vjezbe_Primjeri/airbnb_new_york_city_reviews_full.csv', parse_dates=True)
reviews_df = pd.read_csv('../Data/Vjezbe_Primjeri/airbnb_new_york_city_reviews_full.csv', parse_dates=['date'])
reviews_df

listing_id                  id       date  reviewer_id reviewer_name  \
0             2595               17857 2009-11-21        50679          Jean   
1             2595               19176 2009-12-05        53267          Cate   
2             2595               19760 2009-12-10        38960         Anita   
3             2595               34320 2010-04-09        71130       Kai-Uwe   
4             2595               46312 2010-05-25       117113        Alicia   
...            ...                 ...        ...          ...           ...   
839316    51333702  419451978146866472 2021-08-01    389952287        Latyna   
839317    51337315  418916185474819186 2021-07-31    346570573           Vai   
839318    51344110  420170792523838346 2021-08-02      8178169        Lauren   
839319    51344110  420918895863584518 2021-08-03    347454973       Eli Zoe   
839320    51344110  421704097790531569 2021-08-04     36040182        Arnaud   

                                                 comments  
0       Notre séjour de trois nuits.\r<br/>Nous avons ...  
1                                       Great experience.  
2       I've stayed with my friend at the Midtown Cast...  
3       We've been staying here for about 9 nights, en...  
4       We had a wonderful stay at Jennifer's charming...  
...                                                   ...  
839316                                   Very lovely home  
839317  The host canceled this reservation the day bef...  
839318  Christopher's place is really well located. It...  
839319  Nice place , comfy, all what you need for nice...  
839320  Host has been very kind and responsive. <br/>T...  

[839321 rows x 6 columns]

Imamo šest kolona i 839321 redaka.

Provjerimo koji su tipovi podataka u ovih 6 kolona.

In [190]:
reviews_df.dtypes

listing_id                int64
id                        int64
date             datetime64[ns]
reviewer_id               int64
reviewer_name            object
comments                 object
dtype: object

Provjerom tipa podataka po kolonama, vidimo da kolona za datum ima ispravan tip podatka tako da nema potrebe za njenom konverzijom. Nakon provjere imamo li praznih ćelija, trebat ćemo samo *počistiti* tekstualne podatke u kolonama "reviewer_name" i "comments".

Provjerimo imamo li praznih ćelija i koliko ih je.

In [191]:
reviews_df.isnull().sum()

listing_id         0
id                 0
date               0
reviewer_id        0
reviewer_name      7
comments         812
dtype: int64

Ćelije bez podataka imamo u kolonama "comments" i "reviewer_name". Zato ćemo prvo popuniti te ćelije te nakon toga "počistiti" tekstualne podatke.

In [192]:
reviews_df['comments'].fillna('No Comments for Listing', inplace=True)
reviews_df['reviewer_name'].fillna('J. Doe', inplace=True)

In [193]:
reviews_df.isnull().sum()

listing_id       0
id               0
date             0
reviewer_id      0
reviewer_name    0
comments         0
dtype: int64

Sada više nemamo praznih ćelija. Vrijeme je da *očistimo* tekstualne podatke.

In [194]:
reviews_df['reviewer_name'] = reviews_df['reviewer_name'].str.strip()
reviews_df['comments'] = reviews_df['comments'].str.strip()

In [195]:
reviews_df.isnull().sum()

listing_id       0
id               0
date             0
reviewer_id      0
reviewer_name    0
comments         0
dtype: int64

In [196]:
reviews_df.dtypes

listing_id                int64
id                        int64
date             datetime64[ns]
reviewer_id               int64
reviewer_name            object
comments                 object
dtype: object

**VJEŽBA**<br>
**Napravite skupove podataka s ocjenama korisnika trgovačkog centra te ocjenama učenika.**

In [197]:
mall_customers_df = pd.read_csv('../Data/Vjezbe_Primjeri/mall_customers.csv')
mall_customers_df

CustomerID  Gender  Age  Annual Income (k$)  Spending Score (1-100)
0             1    Male   19                  15                      39
1             2    Male   21                  15                      81
2             3  Female   20                  16                       6
3             4  Female   23                  16                      77
4             5  Female   31                  17                      40
..          ...     ...  ...                 ...                     ...
195         196  Female   35                 120                      79
196         197  Female   45                 126                      28
197         198    Male   32                 126                      74
198         199    Male   32                 137                      18
199         200    Male   30                 137                      83

[200 rows x 5 columns]

In [198]:
ocjene_df = pd.read_csv('../Data/Vjezbe_Primjeri/ocjene.csv')
ocjene_df

fname     lname  gender  age  exercise  hours  grade  \
0        Marcia      Pugh  female   17         3     10   82.4   
1        Kadeem  Morrison    male   18         4      4   78.2   
2          Nash    Powell    male   18         5      9   79.3   
3       Noelani    Wagner  female   14         2      7   83.2   
4       Noelani    Cherry  female   18         4     15   87.4   
...         ...       ...     ...  ...       ...    ...    ...   
1995       Cody  Shepherd    male   19         1      8   80.1   
1996  Geraldine  Peterson  female   16         4     18  100.0   
1997   Mercedes      Leon  female   18         3     14   84.9   
1998     Lucius   Rowland    male   16         1      7   69.1   
1999      Linus    Morris    male   19         4     10   79.6   

                                          address  
0         9253 Richardson Road, Matawan, NJ 07747  
1                33 Spring Dr., Taunton, MA 02780  
2                41 Hill Avenue, Mentor, OH 44060  
3              8839 Marshall St., Miami, FL 33125  
4        8304 Charles Rd., Lewis Center, OH 43035  
...                                           ...  
1995        982 West Street, Alexandria, VA 22304  
1996   78 Morris Street, East Northport, NY 11731  
1997        30 Glenridge Rd., Bountiful, UT 84010  
1998  342 West Meadowbrook Lane, Helena, MT 59601  
1999       81 Homestead Drive, Voorhees, NJ 08043  

[2000 rows x 8 columns]

Vjerujem da ste svi došli do ovoga rješenja rješavajući zadatak iz uvodnog dijela o Pandas biblioteci (osim dijela o popunjavanju praznih ćelija). Ako niste, nije problem, tijekom ovog ponavljanja trebali ste doći na dio koji ćemo koristiti u nastavku.

## Manipulacija podacima - traženje informacija u podacima

Sada kada smo naše podatke pripremili možemo se "igrati" podacima.

Na početku ćemo napraviti nekoliko koraka koji bi mogli biti uključeni u pripremu podataka, ali i manipulaciju podacima. U svakom slučaju, to su koraci koji mogu pomoći u procesu "traženja informacija u podacima".

Prvi takav korak je uklanjanje "viška" kolona. Često skupovi podataka imaju jako puno kolona pa se onda teško usredotočiti na neku skupinu kolona koje nam mogu dati tražene informacije. Osim toga, ponekad postoje kolone koje je sustav generirao tijekom prikupljanja podataka, ali isključivo kao potporu sustavu tijekom procesa prikpljanja podataka. Takve kolone ne predstavljaju nikakvu korist tijekom obrade podataka pa ih je onda korisno maknuti iz skupa podataka.

Metoda koju ćemo koristiti je *drop()* (detalji o metodi iz Pandas dokumentacije: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html?highlight=drop#pandas.DataFrame.drop).

Ova metoda može se koristiti za brisanje, ne samo kolona, nego i redaka. Prvi argument je lista naziva kolone ili kolona koje želimo izbrisati. Naziv argumenta je "labels". Uz ovaj argument,potrebno je navesti i osi (axis) unutar kojih želimo brisati podatke.

*Primjer: dataFrame.drop('naziv_kolone', axis=1).*

Iz gornjeg primjera, vidimo da će se izbrisati element čiji smo naziv proslijedili, a da radi se o koloni koju definira os 1 (axis=1).

Alternativu gornjem primjeru možemo dobiti korištenjem argumenta *columns=\['naziv_kolone', ...\] ili index=\['naziv_indeksa', ...\].*

**Obratite pažnju na to da indeksi NE moraju nužno biti numerički, nego mogu biti i drugačijih zapisa.**

In [199]:
listings_df.drop(['latitude', 'longitude'], axis=1)

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                    Large Furnished Room Near B'way      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood  \
0            Jennifer           Manhattan             Midtown   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant   
2               Garon            Brooklyn  Bedford-Stuyvesant   
3             Rebecca            Brooklyn         Sunset Park   
4            Shunichi           Manhattan             Midtown   
...               ...                 ...                 ...   
36719             Asi           Manhattan             Midtown   
36720      Dina Marie           Manhattan     Upper West Side   
36721  Stay With Vibe           Manhattan     Upper East Side   
36722           Chase              Queens             Arverne   
36723            Nina           Manhattan         Murray Hill   

             room_type  price  minimum_nights  number_of_reviews last_review  \
0      Entire home/apt  150.0              30                 48  2019-11-04   
1      Entire home/apt   75.0               1                408  2021-06-29   
2         Private room   60.0              30                 50  2016-06-05   
3      Entire home/apt  275.0               5                  1  2014-01-02   
4         Private room   61.0               2                485  2021-07-18   
...                ...    ...             ...                ...         ...   
36719  Entire home/apt  255.0              31                  0  1970-01-01   
36720  Entire home/apt  306.0               5                  0  1970-01-01   
36721  Entire home/apt   63.0              30                  0  1970-01-01   
36722     Private room   75.0               1                  0  1970-01-01   
36723  Entire home/apt  115.0              30                  0  1970-01-01   

       reviews_per_month  calculated_host_listings_count  availability_365  \
0                   0.34                               3               341   
1                   5.09                               1               212   
2                   0.55                               1               365   
3                   0.01                               1               184   
4                   3.63                               1               255   
...                  ...                             ...               ...   
36719               0.00                               6               364   
36720               0.00                               1                16   
36721               0.00                              96               334   
36722               0.00                               1                87   
36723               0.00                              11               364   

       last_review_year  minimal_revenue  
0                  2019           4500.0  
1                  2021             75.0  
2                  2016           1800.0  
3                  2014           1375.0  
4                  2021            122.0  
...                 ...              ...  
36719                 0           7905.0  
36720                 0      

**VJEŽBA: uklonite kolone: *'id', 'name', 'host_id', 'host_name'***.

In [200]:
listings_df.drop(['id', 'name', 'host_id', 'host_name'], axis=1)

neighbourhood_group       neighbourhood   latitude  longitude  \
0               Manhattan             Midtown  40.753560 -73.985590   
1                Brooklyn  Bedford-Stuyvesant  40.684940 -73.957650   
2                Brooklyn  Bedford-Stuyvesant  40.685350 -73.955120   
3                Brooklyn         Sunset Park  40.662650 -73.994540   
4               Manhattan             Midtown  40.764570 -73.983170   
...                   ...                 ...        ...        ...   
36719           Manhattan             Midtown  40.755787 -73.965126   
36720           Manhattan     Upper West Side  40.787029 -73.969239   
36721           Manhattan     Upper East Side  40.768950 -73.960455   
36722              Queens             Arverne  40.599257 -73.797953   
36723           Manhattan         Murray Hill  40.748655 -73.981209   

             room_type  price  minimum_nights  number_of_reviews last_review  \
0      Entire home/apt  150.0              30                 48  2019-11-04   
1      Entire home/apt   75.0               1                408  2021-06-29   
2         Private room   60.0              30                 50  2016-06-05   
3      Entire home/apt  275.0               5                  1  2014-01-02   
4         Private room   61.0               2                485  2021-07-18   
...                ...    ...             ...                ...         ...   
36719  Entire home/apt  255.0              31                  0  1970-01-01   
36720  Entire home/apt  306.0               5                  0  1970-01-01   
36721  Entire home/apt   63.0              30                  0  1970-01-01   
36722     Private room   75.0               1                  0  1970-01-01   
36723  Entire home/apt  115.0              30                  0  1970-01-01   

       reviews_per_month  calculated_host_listings_count  availability_365  \
0                   0.34                               3               341   
1                   5.09                               1               212   
2                   0.55                               1               365   
3                   0.01                               1               184   
4                   3.63                               1               255   
...                  ...                             ...               ...   
36719               0.00                               6               364   
36720               0.00                               1                16   
36721               0.00                              96               334   
36722               0.00                               1                87   
36723               0.00                              11               364   

       last_review_year  minimal_revenue  
0                  2019           4500.0  
1                  2021             75.0  
2                  2016           1800.0  
3                  2014           1375.0  
4                  2021            122.0  
...                 ...              ...  
36719                 0           7905.0  
36720                 0           1530.0  
36721                 0           1890.0  
36722                 0             75.0  
36723                 0           3450.0  

[36724 rows x 14 columns]

**PITANJE **<br>
**U tablici opet imamo kolone 'latitude', 'longitude'. To su kolone koje smo maknuli.  Kako ćemo to riješiti?**

ODGOVOR: koristit ćemo *Argument inplace - bool, default False -> inplace=True.*

Maknimo "trajno" kolone.

In [201]:
listings_df.drop(['id', 'name', 'host_id', 'host_name', 'latitude', 'longitude'], axis=1, inplace=True)

Provjerimo jesmo li sada uklonili željene kolone.

In [202]:
listings_df.columns

Index(['neighbourhood_group', 'neighbourhood', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365', 'last_review_year', 'minimal_revenue'],
      dtype='object')

**Promjena naziva kolona**

Ako dohvatimo listu kolona unutar skupa podataka "ocjene_df" (*ocjene_df.columns*) dobit ćemo ovaj rezultat: *Index(\['fname', 'lname', 'gender', 'age', 'exercise', 'hours', 'grade', 'address'\], dtype='object')*.

- 'fname' - predstavlja ime učenika/ce
- 'lname' - predstavlja prezime učenika/ce
- 'gender' - predstavlja spol učenika/ce
- 'age' - predstavlja dob učenika/ce
- 'exercise' - predstavlja tjedni broj sati vježbanja učenika/ce
- 'hours' - predstavlja tjedni broj sati učenja učenika/ce
- 'grade'- predstavlja ocjenu učenika/ce
- 'address'- predstavlja adresu učenika/ce

Za ovaj skup podataka počeli smo koristiti hrvatske nazive pa prevedimo onda i nazive kolona u skladu s gore navedenim opisima.

Promjenu naziva kolone možemo napraviti na više načina.

**Način 1.**

Pomoću atributa columns koji ima svaki *DataFrame* objekt. Jednostavno u listu kolona navedemo nove vrijednosti i to je to:
*ocjene_df.columns = \['Ime', 'Prezime', 'Spol', 'Dob', 'Sati_Tjednog_Vjezbanja', 'Sati_Tjednog_Ucenja', 'Ocjena', 'Puna_Adresa'\].*

**Način 2.**

Pomoću atributa columns i Python liste.

*Korak 1.*<br>
Kreirajmo listu od naziva kolona.<br>
*kolone = list(ocjene_df.columns.values)*

*Korak 2.*<br>
Unesemo željene izmjene.<br>
*kolone\[0\] = 'Ime'*<br>
*kolone\[1\] = 'Prezime'*<br>
*kolone\[2\] = 'Spol'*<br>
*kolone\[3\] = 'Dob'*<br>
*kolone\[4\] = 'Sati_Tjednog_Vjezbanja'*<br>
*kolone\[5\] = 'Sati_Tjednog_Ucenja'*<br>
*kolone\[6\] = 'Ocjena'*<br>
*kolone\[7\] = 'Puna_Adresa'*

*Korak 3.*<br>
Pridružimo vrijednost liste kolekciji kolona skupa podataka.<br>
*ocjene_df.columns = kolone*

**Način 3.**

Pomoću metode *rename()*.
ocjene_df.rename(<br>
&nbsp;&nbsp;&nbsp;&nbsp;columns={<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'fname': 'Ime',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'lname': 'Prezime',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'gender': 'Spol',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'age': 'Dob',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'exercise': 'Sati_Tjednog_Vjezbanja',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'hours': 'Sati_Tjednog_Ucenja',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'grade': 'Ocjena',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'address': 'Puna_Adresa'<br>
&nbsp;&nbsp;&nbsp;&nbsp;}<br>
)<br>

**Zadatak**<br>
**Napravite izmjene pomoću Načina 3. na skupu podataka *ocjene_df.***

**Provjerite nazive kolona za preostala dva skupa podataka te ako mislite da je potrebno, predložite nove nazive kolona.**

In [203]:
ocjene_df.columns = ['Ime', 'Prezime', 'Spol', 'Dob', 'Sati_Tjednog_Vjezbanja', 'Sati_Tjednog_Ucenja', 'Ocjena', 'Puna_Adresa']
ocjene_df

Ime   Prezime    Spol  Dob  Sati_Tjednog_Vjezbanja  \
0        Marcia      Pugh  female   17                       3   
1        Kadeem  Morrison    male   18                       4   
2          Nash    Powell    male   18                       5   
3       Noelani    Wagner  female   14                       2   
4       Noelani    Cherry  female   18                       4   
...         ...       ...     ...  ...                     ...   
1995       Cody  Shepherd    male   19                       1   
1996  Geraldine  Peterson  female   16                       4   
1997   Mercedes      Leon  female   18                       3   
1998     Lucius   Rowland    male   16                       1   
1999      Linus    Morris    male   19                       4   

      Sati_Tjednog_Ucenja  Ocjena                                  Puna_Adresa  
0                      10    82.4      9253 Richardson Road, Matawan, NJ 07747  
1                       4    78.2             33 Spring Dr., Taunton, MA 02780  
2                       9    79.3             41 Hill Avenue, Mentor, OH 44060  
3                       7    83.2           8839 Marshall St., Miami, FL 33125  
4                      15    87.4     8304 Charles Rd., Lewis Center, OH 43035  
...                   ...     ...                                          ...  
1995                    8    80.1        982 West Street, Alexandria, VA 22304  
1996                   18   100.0   78 Morris Street, East Northport, NY 11731  
1997                   14    84.9        30 Glenridge Rd., Bountiful, UT 84010  
1998                    7    69.1  342 West Meadowbrook Lane, Helena, MT 59601  
1999                   10    79.6       81 Homestead Drive, Voorhees, NJ 08043  

[2000 rows x 8 columns]

Ponekad trebamo dodatno prilagoditi tekst naziva kolona tako što će nam trebati sva mala ili sva velika slova u nazivu kolona. Jedan od primjera je potreba za kreiranjem mape za svaku kolonu pa je za Linux OS prikladno koristiti sva mala slova.

Znamo da su kolone unutar DataFrame objekta grupirane unutar liste pa iskoristimo Python za prolazak kroz svaki element liste i primjenimo odgovarajuću metodu nad tekstom naziva kolone.

**Način 1.**

Pomoću Python *map()* funkcije<br>
*ocjene_df.columns = map(str.lower, ocjene_df.columns)*

**Način 2.**
Pomoću Python *for* petlje.<br>
*ocjene_df.columns = \[x.lower() for x in ocjene_df.columns\]*

**Zadatak**<br>
**Napravite izmjene pomoću Načina 2. na skupu podataka *ocjene_df.***

**Na preostala dva skupa podataka prebacite nazive kolona u velika slova.**

In [204]:
#ocjene_df.columns = map(str.lower, ocjene_df.columns)
ocjene_df.columns = [x.lower() for x in ocjene_df.columns]
ocjene_df

ime   prezime    spol  dob  sati_tjednog_vjezbanja  \
0        Marcia      Pugh  female   17                       3   
1        Kadeem  Morrison    male   18                       4   
2          Nash    Powell    male   18                       5   
3       Noelani    Wagner  female   14                       2   
4       Noelani    Cherry  female   18                       4   
...         ...       ...     ...  ...                     ...   
1995       Cody  Shepherd    male   19                       1   
1996  Geraldine  Peterson  female   16                       4   
1997   Mercedes      Leon  female   18                       3   
1998     Lucius   Rowland    male   16                       1   
1999      Linus    Morris    male   19                       4   

      sati_tjednog_ucenja  ocjena                                  puna_adresa  
0                      10    82.4      9253 Richardson Road, Matawan, NJ 07747  
1                       4    78.2             33 Spring Dr., Taunton, MA 02780  
2                       9    79.3             41 Hill Avenue, Mentor, OH 44060  
3                       7    83.2           8839 Marshall St., Miami, FL 33125  
4                      15    87.4     8304 Charles Rd., Lewis Center, OH 43035  
...                   ...     ...                                          ...  
1995                    8    80.1        982 West Street, Alexandria, VA 22304  
1996                   18   100.0   78 Morris Street, East Northport, NY 11731  
1997                   14    84.9        30 Glenridge Rd., Bountiful, UT 84010  
1998                    7    69.1  342 West Meadowbrook Lane, Helena, MT 59601  
1999                   10    79.6       81 Homestead Drive, Voorhees, NJ 08043  

[2000 rows x 8 columns]

**Filtriranje podataka**

Ponekad želimo izdvojiti ili dohvatiti određene podatke iz grupe. Ovisno o cilju, najčešće ćemo definirati neke uvjete. Kao i do sada, ovo možemo napraviti na više načina.

Započnimo korištenjem *str.contains()* metodom. Dohvatimo sve podatke o nekretninama iz Manhattana. Prvo ćemo provjeriti ima li bilo koji podatak koji zadovoljava ovaj uvjet:

In [205]:
listings_df['neighbourhood_group'].str.contains('Manhattan').any()

True

Rezultat je *True*, dakle postoje takvi podaci u skupu podataka. Provjerimo jesu li svi podaci s ovim uvjetom.

In [206]:
listings_df['neighbourhood_group'].str.contains('Manhattan').all()

False

Za ovaj skup podataka je očekivano, ali je ponekad korisno provjeriti jer ako svi podaci zadovoljavaju određeni uvjet, onda nema potrebe za filtriranjem. Probajmo dohvatiti u kojim redovima su podaci koji ispunjavaju traženi uvjet.

In [207]:
listings_df['neighbourhood_group'].str.contains('Manhattan')

0         True
1        False
2        False
3        False
4         True
         ...  
36719     True
36720     True
36721     True
36722    False
36723     True
Name: neighbourhood_group, Length: 36724, dtype: bool

Dobili smo samo podatak ispunjava li redak uvjet ili ne. Međutim, za dohvat svih podataka u retku, trebamo se koristiti svojstvom *loc\[\]* koje kao argument prima listu Boolean vrijednosti. Napravimo to, a kao vrijednost argumenta  predat ćemo gornju liniju koda te dodati uvjet da je jednako *True*, kako nam ne bi vratio sve.

In [208]:
listings_df.loc[listings_df['neighbourhood_group'].str.contains('Manhattan') == True]

neighbourhood_group    neighbourhood        room_type  price  \
0               Manhattan          Midtown  Entire home/apt  150.0   
4               Manhattan          Midtown     Private room   61.0   
5               Manhattan  Upper West Side     Private room   75.0   
8               Manhattan      East Harlem     Private room   65.0   
9               Manhattan      East Harlem     Private room   62.0   
...                   ...              ...              ...    ...   
36718           Manhattan  Upper East Side  Entire home/apt  125.0   
36719           Manhattan          Midtown  Entire home/apt  255.0   
36720           Manhattan  Upper West Side  Entire home/apt  306.0   
36721           Manhattan  Upper East Side  Entire home/apt   63.0   
36723           Manhattan      Murray Hill  Entire home/apt  115.0   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
0                  30                 48  2019-11-04               0.34   
4                   2                485  2021-07-18               3.63   
5                   2                118  2017-07-12               0.90   
8                  30                  0  1970-01-01               0.00   
9                  30                234  2021-07-04               1.88   
...               ...                ...         ...                ...   
36718              30                  0  1970-01-01               0.00   
36719              31                  0  1970-01-01               0.00   
36720               5                  0  1970-01-01               0.00   
36721              30                  0  1970-01-01               0.00   
36723              30                  0  1970-01-01               0.00   

       calculated_host_listings_count  availability_365  last_review_year  \
0                                   3               341              2019   
4                                   1               255              2021   
5                                   1                 0              2017   
8                                   2               365                 0   
9                                   1               258              2021   
...                               ...               ...               ...   
36718                              96               335                 0   
36719                               6               364                 0   
36720                               1                16                 0   
36721                              96               334                 0   
36723                              11               364                 0   

       minimal_revenue  
0               4500.0  
4                122.0  
5                150.0  
8               1950.0  
9               1860.0  
...                ...  
36718           3750.0  
36719           7905.0  
36720           1530.0  
36721           1890.0  
36723           3450.0  

[16105 rows x 12 columns]

**Zadatak**<br>
**Iz skupa podataka o ocjenama, prikažite samo retke učenika s ocjenama većim od 70.
Dodajte filter i po spolu.**<br>
**VAŽNO: koristit ćemo znakove za logičke operacije iz C programskog jezika:**
- **& - and**
- **| - or.**

In [209]:
ocjene_df.loc[(ocjene_df['ocjena'] > 70) & (ocjene_df['spol'] == 'female')]

ime   prezime    spol  dob  sati_tjednog_vjezbanja  \
0        Marcia      Pugh  female   17                       3   
3       Noelani    Wagner  female   14                       2   
4       Noelani    Cherry  female   18                       4   
6         Nelle    Golden  female   17                       1   
8       Illiana     Rojas  female   15                       5   
...         ...       ...     ...  ...                     ...   
1990      Adena    Battle  female   17                       2   
1992   Isabelle    Barber  female   14                       5   
1993       Risa    Watson  female   14                       2   
1996  Geraldine  Peterson  female   16                       4   
1997   Mercedes      Leon  female   18                       3   

      sati_tjednog_ucenja  ocjena  \
0                      10    82.4   
3                       7    83.2   
4                      15    87.4   
6                       9    80.2   
8                       9    76.5   
...                   ...     ...   
1990                    8    70.2   
1992                    9    78.5   
1993                   10    74.3   
1996                   18   100.0   
1997                   14    84.9   

                                        puna_adresa  
0           9253 Richardson Road, Matawan, NJ 07747  
3                8839 Marshall St., Miami, FL 33125  
4          8304 Charles Rd., Lewis Center, OH 43035  
6             9768 Hanover Dr., Meadville, PA 16335  
8          9425 Studebaker Dr., Thibodaux, LA 70301  
...                                             ...  
1990    9272 Elizabeth Drive, Londonderry, NH 03053  
1992      955 Glen Ridge Rd., Plattsburgh, NY 12901  
1993  37 Augusta Lane, Montgomery Village, MD 20886  
1996     78 Morris Street, East Northport, NY 11731  
1997          30 Glenridge Rd., Bountiful, UT 84010  

[901 rows x 8 columns]

**Statistička obrada podataka**

Prošli smo filtriranje podataka i prije nego se upoznamo s grupiranjem, upoznajmo se sa statističkom obradom podataka. Statistika će nam pomoći da se još bolje upoznamo s našim podacima.

Za početak provjerimo najčešće korištene statističke vrijednosti:
- minimalna vrijednost (*min*)
- maksimalna vrijednost (*max*)
- raspon (*max - min*)
- srednja vrijednosti (*mean*)
- medijan (*median*) - vrijednost koja se nalazi u sredini nakon što podatke poredamo po veličini
- standardna devijacija (*std*)
- ukupan broj vrijednosti (*count*) na kolonama koje su nam najzanimljivije za ove podatke.

Za skup podataka o nekretninama to su kolone:
- *price*
- *minimum_nights*
- *number_of_reviews*
- *last_review*
- *reviews_per_month*
- *availability_365*
- *last_review_year*
- *minimal_revenue*

*INFORMACIJA: kada budemo koristili vizualizaciju podataka, onda ćemo ove vrijednosti prikazati grafički pa će biti puno lakše uočiti neke granične vrijednosti.*

In [210]:
# price kolona
# INFO ":>10" NIJE provjera veće ili manje nego poravnanje desno za 10 znakova
print("CIJENA")
print('-' * 50)
print(f"Naziv vrijednosti\t\t\tVrijednost")
print('-' * 50)
print(f"Minimalna cijena:\t\t{listings_df['price'].min():>10} USD")
print(f"Maksimalna cijena:\t\t{listings_df['price'].max():>10} USD")
print(f"Raspon cijena:\t\t\t{(listings_df['price'].max() - listings_df['price'].min()):>10} USD")
print(f"Prosječna cijena:\t\t{round(listings_df['price'].mean(), 2):>10} USD")
print(f"Medijan cijene:\t\t\t{round(listings_df['price'].median(), 2):>10} USD")
print(f"Standardna devijacija cijene:\t{round(listings_df['price'].std(), 2):>10} USD")
print(f"Ukupan broj vrijednosti cijena:\t{round(listings_df['price'].count(), 2):>10}")
print('-' * 50)
print('\n')

CIJENA
--------------------------------------------------
Naziv vrijednosti			Vrijednost
--------------------------------------------------
Minimalna cijena:		       0.0 USD
Maksimalna cijena:		   10000.0 USD
Raspon cijena:			   10000.0 USD
Prosječna cijena:		    160.11 USD
Medijan cijene:			     108.0 USD
Standardna devijacija cijene:	    293.89 USD
Ukupan broj vrijednosti cijena:	     36724
--------------------------------------------------




**VJEŽBA**<br>
**Dovršite analizu za preostale kolone.**

Srećom, Pandas ima ugrađenu funkciju *describe()* koja objedinjuje većinu gore navedene statističke vrijednosti.

In [211]:
listings_df.describe()

price  minimum_nights  number_of_reviews  reviews_per_month  \
count  36724.000000    36724.000000       36724.000000       36724.000000   
mean     160.109057       22.482083          22.854836           1.022812   
std      293.888431       28.361534          49.188479           3.374019   
min        0.000000        1.000000           0.000000           0.000000   
25%       67.000000        3.000000           0.000000           0.000000   
50%      108.000000       30.000000           4.000000           0.170000   
75%      175.000000       30.000000          19.000000           0.990000   
max    10000.000000     1250.000000         998.000000         192.930000   

       calculated_host_listings_count  availability_365  last_review_year  \
count                    36724.000000      36724.000000      36724.000000   
mean                        13.432306        122.570608       1501.684049   
std                         46.293004        138.765815        881.743887   
min                          1.000000          0.000000          0.000000   
25%                          1.000000          0.000000          0.000000   
50%                          1.000000         57.000000       2019.000000   
75%                          3.000000        259.000000       2021.000000   
max                        335.000000        365.000000       2021.000000   

       minimal_revenue  
count     3.672400e+04  
mean      3.562474e+03  
std       1.374511e+04  
min       0.000000e+00  
25%       4.600000e+02  
50%       1.860000e+03  
75%       3.900000e+03  
max       1.170000e+06

Možemo dodati i argument *include='all'* koji uključuje sve dostupne izračune.

In [212]:
listings_df.describe(include='all')

<ipython-input-212-b8fd02d50633>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  listings_df.describe(include='all')


neighbourhood_group       neighbourhood        room_type         price  \
count                36724               36724            36724  36724.000000   
unique                   5                 223                4           NaN   
top              Manhattan  Bedford-Stuyvesant  Entire home/apt           NaN   
freq                 16105                2690            19484           NaN   
first                  NaN                 NaN              NaN           NaN   
last                   NaN                 NaN              NaN           NaN   
mean                   NaN                 NaN              NaN    160.109057   
std                    NaN                 NaN              NaN    293.888431   
min                    NaN                 NaN              NaN      0.000000   
25%                    NaN                 NaN              NaN     67.000000   
50%                    NaN                 NaN              NaN    108.000000   
75%                    NaN                 NaN              NaN    175.000000   
max                    NaN                 NaN              NaN  10000.000000   

        minimum_nights  number_of_reviews          last_review  \
count     36724.000000       36724.000000                36724   
unique             NaN                NaN                 2452   
top                NaN                NaN  1970-01-01 00:00:00   
freq               NaN                NaN                 9415   
first              NaN                NaN  1970-01-01 00:00:00   
last               NaN                NaN  2021-08-04 00:00:00   
mean         22.482083          22.854836                  NaN   
std          28.361534          49.188479                  NaN   
min           1.000000           0.000000                  NaN   
25%           3.000000           0.000000                  NaN   
50%          30.000000           4.000000                  NaN   
75%          30.000000          19.000000                  NaN   
max        1250.000000         998.000000                  NaN   

        reviews_per_month  calculated_host_listings_count  availability_365  \
count        36724.000000                    36724.000000      36724.000000   
unique                NaN                             NaN               NaN   
top                   NaN                             NaN               NaN   
freq                  NaN                             NaN               NaN   
first                 NaN                             NaN               NaN   
last                  NaN                             NaN               NaN   
mean             1.022812                       13.432306        122.570608   
std              3.374019                       46.293004        138.765815   
min              0.000000                        1.000000          0.000000   
25%              0.000000                        1.000000          0.000000   
50%              0.170000                        1.000000         57.000000   
75%              0.990000                        3.000000        259.000000   
max            192.930000                      335.000000        365.000000   

        last_review_year  minimal_revenue  
count       36724.000000     3.672400e+04  
unique               NaN              NaN  
top                  NaN              NaN  
freq                 NaN              NaN  
first                NaN              NaN  
last                 NaN              NaN  
mean         1501.684049     3.562474e+03  
std           881.743887     1.374511e+04  
min             0.000000     0.000000e+00  
25%             0.000000     4.600000e+02  
50%          2019.000000     1.860000e+03  
75%          2021.000000     3.900000e+03  
max          2021.000000     1.170000e+06

*INFORMACIJA: dodajte traženi argument.*

Gore navedeni statistički izračuni vrijede za sve kolone ili za sve retke unutar neke kolone (serije). Primjenimo stečeno znanje o statističkim izračunima na ono što smo već naučili o filtriranju podataka.

In [213]:
listings_df.loc[listings_df['price'] > 2600].mean()

# listings_df.loc[listings_df['price'] > 1000].mean()
# round(listings_df.loc[listings_df['price'] > 1000].mean(), 2)

<ipython-input-213-f06bb222f1a8>:1: FutureWarning: DataFrame.mean and DataFrame.median with numeric_only=None will include datetime64 and datetime64tz columns in a future version.
  listings_df.loc[listings_df['price'] > 2600].mean()


price                               5703.890909
minimum_nights                        28.727273
number_of_reviews                      5.800000
reviews_per_month                      0.281818
calculated_host_listings_count         2.400000
availability_365                     169.345455
last_review_year                     954.254545
minimal_revenue                   177682.109091
dtype: float64

**Sortiranje i grupiranje podataka**

Često podatke trebamo sortirati po veličini, od najmanje prema najvećoj ili obrnuto. To radimo pomoću metode *sort_values()* koja ima dva najvažnija parametra:
- *by* - listu kolona koje želimo sortirati i
- *ascending* - koji određuje je li sortiranje od manje prema većoj vrijednosti ili obrnuto.

In [214]:
ocjene_df.sort_values(by='dob', ascending=False)
# ocjene_df.sort_values(by='dob', ascending=True)

ime    prezime    spol  dob  sati_tjednog_vjezbanja  \
1000    Hanna     Mooney  female   19                       2   
902    Steven     Graham    male   19                       4   
818   Colette      Rojas  female   19                       4   
822     Rinah   Jacobson  female   19                       1   
826     Hayes  Wilkinson    male   19                       5   
...       ...        ...     ...  ...                     ...   
85     Porter       Ware    male   14                       1   
587     Kiona    Collier  female   14                       4   
588   Jocelyn     Ashley  female   14                       3   
89     Lucius    Herrera    male   14                       2   
294   Illiana     Branch  female   14                       5   

      sati_tjednog_ucenja  ocjena  \
1000                   10    88.8   
902                    17    95.1   
818                     6    73.3   
822                     7    80.8   
826                     3    76.0   
...                   ...     ...   
85                     16    90.4   
587                    11    84.0   
588                     9    73.4   
89                     19    88.0   
294                     4    69.8   

                                       puna_adresa  
1000    8293 SW. Cedar Swamp Lane, Union, NJ 07083  
902                  90 Grove St., Camas, WA 98607  
818        17 Pin Oak Street, Villa Park, IL 60181  
822         737 Amherst Court, Amsterdam, NY 12010  
826         350 Temple Court, Mason City, IA 50401  
...                                            ...  
85             52 Shirley Drive, Merrick, NY 11566  
587   827 Rockledge Rd., New Port Richey, FL 34653  
588         830 Amerige Drive, Davenport, IA 52804  
89              67 Old Clinton Dr., Lynn, MA 01902  
294     9730 Van Dyke Court, Port Orange, FL 32127  

[2000 rows x 8 columns]

Rekli smo da možemo sortirati više kolona. Jednostavno navedimo više kolona kao vrijednost *by* parametra.

In [215]:
ocjene_df.sort_values(by=['dob', 'sati_tjednog_ucenja'], ascending=True)
# ocjene_df.sort_values(by=['dob', 'sati_tjednog_ucenja'], ascending=True).head(50)

ime    prezime    spol  dob  sati_tjednog_vjezbanja  \
1213  Arthur     Nieves    male   14                       3   
1640   Nolan      Small    male   14                       3   
1758    Omar    Pittman    male   14                       5   
1916   Kevin   Townsend    male   14                       5   
44     Dolan       Ford    male   14                       3   
...      ...        ...     ...  ...                     ...   
36      Ivor     Arnold    male   19                       4   
101     Sage  Cleveland  female   19                       2   
204    Timon    Goodman    male   19                       4   
508      Dai    Osborne  female   19                       5   
954   Athena     Harvey  female   19                       5   

      sati_tjednog_ucenja  ocjena                                 puna_adresa  
1213                    0    76.8    377 Beech Street, Chesterfield, VA 23832  
1640                    2    71.5   9205 Myers Drive, Ellicott City, MD 21042  
1758                    2    66.6     281 Tower Street, Saint Johns, FL 32259  
1916                    2    60.1  9805 Walnutwood Dr., Panama City, FL 32404  
44                      3    64.2                12 New St., Newark, NJ 07103  
...                   ...     ...                                         ...  
36                     20   100.0    7027 Magnolia Dr., Catonsville, MD 21228  
101                    20   100.0         9721B Green Dr., Fairhope, AL 36532  
204                    20    98.0  3 North Rockaway Ave., Uniontown, PA 15401  
508                    20   100.0        2 Pilgrim Road, Alexandria, VA 22304  
954                    20   100.0    137 Cedar Swamp Street, Laurel, MD 20707  

[2000 rows x 8 columns]

**ZADATAK**<br>
**Sortirajte jednu ili više kolona iz grupa podataka o nekretninama i njihovim ocjenama (*listings_df*, *reviews_df*).**

**Grupiranje podataka**

Podatke često grupiramo po nekoj kategoriji. Na primjer, svi apartmani ili svi učenici neke dobi. Za tu grupu podataka onda napravimo određeni statistički izračun (srednja vrijednost, medijan i sl.).

Iz primjera niže, prvo smo izdvojili tri kolone, od kojih smo onda napravili grupiranje po tipu nekretnine i za svaku od tih grupa smo izračunali srednju vrijednost.

*INFORMACIJA: ukoliko ste "izgubili" kolonu 'minimal_revenue', a može se dogoditi jer radimo u jednom velikom Python programu, ponovite njezino kreiranje. Ovo se može dogoditi zato što radimo dosta primjera, često pokrećemo različite ćelije i to ne nužno onim redoslijedom kako smo ih napisali pa zbog toga može doći do grešaka. U stvarnom životu, kada radite s jednom skupinom podataka, odmah je razdvojite na manje cjeline i onda na svakoj manjoj cjelini radite zasebno.*

In [225]:
# listings_df['minimal_revenue'] = listings_df['price'] * listings_df['minimum_nights']
listings_df[['room_type', 'price', 'minimal_revenue']].groupby('room_type', as_index=False).mean()

room_type       price  minimal_revenue
0  Entire home/apt  213.324779      4949.334480
1       Hotel room  293.262548      2662.552124
2     Private room   96.617690      1963.337077
3      Shared room  107.978297      2575.096828

Podatke možemo grupirati, ne samo u jednoj koloni, nego u više njih.

In [226]:
listings_df[['room_type', 'last_review_year', 'price']].groupby(['room_type', 'last_review_year'], as_index=False).median()

room_type  last_review_year  price
0   Entire home/apt                 0  180.0
1   Entire home/apt              2010  228.0
2   Entire home/apt              2011  165.0
3   Entire home/apt              2012  150.0
4   Entire home/apt              2013  150.0
5   Entire home/apt              2014  170.0
6   Entire home/apt              2015  150.0
7   Entire home/apt              2016  150.0
8   Entire home/apt              2017  149.0
9   Entire home/apt              2018  150.0
10  Entire home/apt              2019  150.0
11  Entire home/apt              2020  148.0
12  Entire home/apt              2021  150.0
13       Hotel room                 0  370.0
14       Hotel room              2016  349.0
15       Hotel room              2017  248.0
16       Hotel room              2018  125.0
17       Hotel room              2019  291.0
18       Hotel room              2020  211.5
19       Hotel room              2021  165.0
20     Private room                 0   68.0
21     Private room              2011  349.0
22     Private room              2012   90.0
23     Private room              2013  100.0
24     Private room              2014   80.0
25     Private room              2015   75.0
26     Private room              2016   65.0
27     Private room              2017   65.0
28     Private room              2018   65.0
29     Private room              2019   69.0
30     Private room              2020   63.0
31     Private room              2021   68.0
32      Shared room                 0   65.0
33      Shared room              2013   50.0
34      Shared room              2014   82.0
35      Shared room              2015   61.5
36      Shared room              2016   95.0
37      Shared room              2017   51.0
38      Shared room              2018   57.0
39      Shared room              2019   55.0
40      Shared room              2020   45.0
41      Shared room              2021   40.0

**ZADATAK**<br>
**Grupirajte jednu ili više kolona iz grupa podataka o ocjenama učenika.**

**Korelacija između podataka u kolonama**

Često ćemo željeti otkriti u kakvoj su vezi podaci iz jedne kolone s podacima iz druge, odnosno postoji li nekakva korelacija između njih. Na primjer, broj sati učenja i ocjena mogu biti u korelaciji, ali ime i ocjena sigurno ne bi trebali imati međusobni utjecaj, odnosno korelaciju.

Pandas ima ugrađenu metodu nad *DataFrame* objektom naziva *corr()*. Ova metoda vraća vrijednosti do 1 do -1. Ako je korelacija između dviju kolona visoka, onda će vrijednost biti 1 ili jako blizu jedinice, a ako kolone nisu u korelaciji, onda će vrijednost biti negativna, odnosno bit će blizu minus jedan.

Pogledajmo kako *corr()* metoda radi na skupu podataka s ocjenama učenika.

In [229]:
ocjene_df.corr()

dob  sati_tjednog_vjezbanja  sati_tjednog_ucenja  \
dob                     1.000000               -0.003643            -0.017467   
sati_tjednog_vjezbanja -0.003643                1.000000             0.021105   
sati_tjednog_ucenja    -0.017467                0.021105             1.000000   
ocjena                 -0.007580                0.161286             0.801955   

                          ocjena  
dob                    -0.007580  
sati_tjednog_vjezbanja  0.161286  
sati_tjednog_ucenja     0.801955  
ocjena                  1.000000

U tabeli se vidi da najveću vrijednost (0.801955) imaju kolone ocjena i *sati_tjednog_ucenja*. Zatim kolone *ocjena* i *sati_tjednog_vjezbanja*.

**ZADATAK**<br>
**Pronađite korelacije za skupine podataka o nekretninama i njihovim ocjenama (*listings_df*, *reviews_df*).**

**PIVOT tabele**

Pivot tabele ste vjerojatno upoznali u Microsoft Excelu. Njihova primjena je sumirati, sortirati, reorganizirati, grupirati, prebrojati sve ili prosjek podataka u tabeli.

Možda ovo pomogne boljem razumijevanju pivot tabele. Pivot tabela može:
- grupirati pojedinačne zapise, redove u kategorije
- prebrojati elemente u svakoj kategoriji
- sumirati vrijednosti elemenata
- izračunati srednju vrijednost, minimum ili maksimum.

Pandas ima ugrađenu metodu *pivot_table()*. Upoznajmo se s njezinim mogućnostima. Više detalja na: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html?highlight=pivot_table#pandas.pivot_table.

In [239]:
pd.pivot_table(ocjene_df, index=['spol'])

dob   ocjena  sati_tjednog_ucenja  sati_tjednog_vjezbanja
spol                                                                
female  16.568  82.7173               10.932                   3.047
male    16.589  82.3948               11.045                   2.954

Srećom, nema puno kolona, ali zamislimo da ima pa ćemo odabrati prikaz samo prosječne ocjene po spolu i dobi.

In [251]:
pd.pivot_table(ocjene_df, values=['ocjena'], index=['spol', 'dob'])

ocjena
spol   dob           
female 14   82.058228
       15   82.905696
       16   82.582278
       17   83.599435
       18   82.500000
       19   82.592216
male   14   82.323418
       15   82.671831
       16   82.752841
       17   82.949721
       18   82.103593
       19   81.598315

Bilo bi dobro kada bismo umjesto prosječne vrijednosti dobili najmanju vrijednost ocjene po spolu.

In [242]:
pd.pivot_table(ocjene_df, values=['ocjena'], index=['spol'], aggfunc='min')

ocjena
spol          
female    32.0
male      43.0

**ZADATAK**<br>
**Očigledno argument *aggfunc* predstavlja vrijednost funkcije po kojoj se agregira, odnosno grupira, odnosno predstavlja statističku vrijednost koju računamo. Probajte mijenjati *aggfunc* argument nazivima statističkih metoda u Pandasu te uz kolonu ocjene, dodajte i sate tjednog učenja (otkrili smo da su u koleraciji).**

Bilo bi korisno kada bismo na kraju *pivot* tabele, kao zadnji redak, imali i sumarni prikaz za sve vrijednosti određene kolone. Recimo, prosječna vrijednost za ocjene i sate tjednog učenja po spolu, ali i ukupna prosječna vrijednosti za ocjene i sate tjednog učenja BEZ obzira na spol.

In [252]:
pd.pivot_table(ocjene_df, values=['ocjena', 'sati_tjednog_ucenja'], index=['spol'], aggfunc='mean', margins=True)

ocjena  sati_tjednog_ucenja
spol                                 
female  82.71730              10.9320
male    82.39480              11.0450
All     82.55605              10.9885

### ZADACI ZA VJEŽBANJE

Analizirajte csv datoteke:

- *cars.csv*
- *e-cars.csv*
- *earthquakes.csv*

tako da pomoću Pandas metoda (*corr(), pivot_table() ...*) izvučete neke zanimljive informacije.